subsys/shell/shell_uart.c

```diff
@@ -278,6 +278,8 @@ static int enable_shell_uart(struct device *arg)
 		return -ENODEV;
 	}
 
+	smp_shell_init();
+
```

Comment:
This makes shell_uart dependent on SMP over shell, which itself is Kconfig configurable (`CONFIG_MCUMGR_SMP_SHELL`) and the `ubsys/mgmt/smp_shell.c` is included into compilation conditionally (`subsys/mgmt/CMakeLists.txt`).
I think that this invocation makes SMP via shell obligatory when shell is enabled, even if not needed by application, which should not happen.

Reply:
The trick is in include/mgmt/smp_shell.h:
```
#ifdef CONFIG_MCUMGR_SMP_SHELL
int smp_shell_init(void);
#else
static inline int smp_shell_init(void)
{
	return -ENOTSUP;
}
#endif
```
implemented similar to `net_ppp_ping` in `include/net/ppp.h`.

The idea was to always export API to user, just like in #26875 .

Reply:
Sorry I missed that. This is misleading actually, because it calls function that by name informs dev that something will happen and actually does nothing. Maybe just ifdef it in place of invocation and just not define the error returning function at all?

Reply:
I think that more ifdefs make code less readable. And by invoking `smp_shell_init()` you ask SMP shell to be initialized. But it returns `-ENOTSUP`, because SMP shell is not enabled in configuration. I think this is similar to `uart_tx()`, which will return `-ENOTSUP` if there is no UART enabled in the system.

Also it is much easier to keep buildable code with different configuration variants, when using static inline dummy functions just like in this PR.

Reply:
I understand you arguments about `ifdefs` but you still use them to define the error returning function.
Why not just do something similar:
```
if (IS_ENABLED(CONFIG_MCUMGR_SMP_SHELL)) {
    smp_shell_init(void)
}
```
Wouldn't that work? It is kind a compromise, it requires prototype of `smp_shell_init(void)` to compile, but compiler will optimize out the code and will not require linking the `smp_shell_init(void)`. No `ifdefs`.

I think that the uart is written this way to enable such `if...else`, where one block serves sync the other async uart. This would enable compilation of both blocks but a compiler would optimize out the one that is not used. In case when someone uses functions for the wrong mode, the returned error codes would allow to detect the problem.

Reply:
With the changes https://github.com/zephyrproject-rtos/zephyr/pull/26426
> I think this is similar to uart_tx(), which will return -ENOTSUP if there is no UART enabled in the system.

it will no longer be the case

Reply:
Changed to use `IS_ENABLED()`, as requested.